### DISCUSSION TITLE
update optional DateTime to now()

### DISCUSSION THREAD

**Discussion by vogler** [prisma/prisma#5582] [Q&A]
Is there some way to update an optional DateTime to `now()`? `@updatedAt` does not work since there's nothing else to update. 
Currently I'm setting it to `new Date()` but that comes from the client which might be not in sync with the server's time.

~~~prisma
model Time {
  todoId Int
  todo   Todo      @relation(fields: [todoId], references: [id])
  start  DateTime  @default(now())
  end    DateTime?
  @@id([todoId, start])
}
~~~

I first create a row with `end = null` and later want to set it to `now()`.
~~~typescript
... create: { } // start = now(), end = null
... updateMany: { data: { end: new Date() }, where: { end: null } } // end should be set to now()
~~~

With `end    DateTime? @updatedAt` and the following, `end` is not updated at all:
~~~typescript
... create: { end: null } // start = now(), end = null
... updateMany: { data: { }, where: { end: null } } // end is not updated :(
~~~

**Comment by ryands17** [comment]
@vogler 👋 

The reason the second snippet with `@updateAt` is not working is because `data` is empty (`{}`) in your `updateMany` query which means that there is nothing to update. `@updatedAt` will only be updated if at least 1 change was made.

I think in this case it's best to go with the 1st option as that would be fine and Prisma always stores the `DateTime` in UTC so synchronisation shouldn't be a problem.

**Reply by vogler** [reply]
>  `@updatedAt` will only be updated if at least 1 change was made.

The problem seems to be that it's strict in null, i.e., even if I update `start`, an `end` with `null` will not be updated.

> I think in this case it's best to go with the 1st option as that would be fine and Prisma always stores the `DateTime` in UTC so synchronisation shouldn't be a problem.

It is a problem whenever the query is created on a different device than the database runs on (e.g. different server).
In my case `start` is the database's time, but `end` is the client's time since the query is built there. Potentially this could break the invariant `start < end`.

**Reply by ryands17** [reply]
> The problem seems to be that it's strict in null, i.e., even if I update start, an end with null will not be updated.

I tried and it does update. I used the following schema:

```prisma
model Time {
  id    Int       @id @default(autoincrement())
  something String?
  start DateTime  @default(now())
  end   DateTime? @updatedAt
}
```

And ran the following query:

```ts
  let time = await prisma.time.create({
    data: { start: new Date(), end: null, something: 'something' },
  })
  console.log(time)

  await sleep()

  time = await prisma.time.update({
    data: { something: 'other' },
    where: { id: time.id },
  })
  console.log(time)
```

The output was with the updated time:

![image](https://user-images.githubusercontent.com/19697099/107624140-73d6c200-6c80-11eb-8904-b92f834c711c.png)

> It is a problem whenever the query is created on a different device than the database runs on (e.g. different server).
In my case start is the database's time, but end is the client's time since the query is built there. Potentially this could break the invariant start < end.

As long as your database time is UTC, this will not happen as Prisma does that conversion automatically.

**Reply by vogler** [reply]
Mh, I don't want to introduce an otherwise useless field for it.

> As long as your database time is UTC, this will not happen as Prisma does that conversion automatically.

I'm not talking about time zones, but different clocks.
In your example, assume the server's clock is 10s ahead of the db server's clock, then `start` > `end`.
If you have control over the servers and they use NTP, that might not be a problem, but in my case the `new Date()` comes from the browser.

A solution would be for `DateTime` to also allow e.g. `now()` as a string literal to pass through to the db.

**Reply by ryands17** [reply]
> I'm not talking about time zones, but different clocks.

In that case, the only option remaining is a [raw query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#executeraw). I would suggest opening a feature request for this [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that it can directly allowed via Prisma's API as this seems like a valid use case.

**Reply by vogler** [reply]
Thanks!
Ref: https://github.com/prisma/prisma/issues/5598

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
